packages/sync/src/conflicts/base/BaseStateLink.ts

```diff
@@ -0,0 +1,31 @@
+
+import { ApolloLink, NextLink, Operation } from "apollo-link";
+import { BaseStateProvider } from "./BaseStateProvider";
+
+/**
+ * Apollo link used to manage state of the objects before 
+ * they are sent to server. ObjectStates are used in conflicts to determine 
+ * difference between server and client data.
+ *
+ * @see BaseStateProvider for more information
+ */
+export class BaseStateLink extends ApolloLink {
+
+  constructor(private baseState: BaseStateProvider) {
+    super();
+  }
+
+  public request(operation: Operation, forward?: NextLink) {
+    // TODO get data from cache
+    // TODO filter only for mutations (EDITS)
+    // TODO Add persistence depending if offline or not
+    this.baseState.save({}, operation.toKey(), false);
```

Comment:
what is the return value for `operation.toKey()`, will it be unique for each version of the base state object? The reason I am asking is thinking about what will happen if when mutation A is in flight and waiting for response from the server, mutation B happened on the same object? Will it override the base state for mutation A if the key is the same?

Reply:
Very valid point! We hav
This is WIP so some of the code elements are still not solid enough.
We have unique keys we generate for offline etc.

The `operation.toKey()` is used temporarily. Key will contain hash, passed variables etc. so it will be unique itself. The problem we facing now is to actually access base data. To access base data we will need to know the actual type we request from server and this is known only when request is finished. This Type is not available in Apollo-client for javascript (it is well known in Android and IOS). To overcome that we can:

- force application wide id's in a cache (nah.. this is to restrictive and limiting for current integrations)
- require type information to be passed when mutating 
This can be done by:
a) Directive - @returnType(name='Task')
b) Simply reusing cache update helpers that currently also require type information
- Another idea.

We trying our best to avoid storing entire history on the server, but there will be number of challenges to overcome and there will be always tradeof. 

To provide more context. We will get type information from server after the successful request but then this is too late.

We could technically access base when conflict happened (after request) but after investigating this case we found out that base can often be overridden by another query, subscription or even previous mutation from offline store so this method will not be 100% reliable and in some cases we will need to fall back to conflict resolution without diff which is not desired. 

Entire PR is still in the progress but once we figure out how to access cache data before sending request to server we will be in a very good position to resolve this problem without any corner cases and fallbacks to non diff resolutions.

Reply:
> require type information to be passed when mutating
> This can be done by:
> a) Directive - @returnType(name='Task')
> b) Simply reusing cache update helpers that currently also require type information

I like that idea. Just looking at it from a slightly different perspective. Could it be possible to include a client side directive, something like `conflictsOn(TYPE: String)`?

The purpose of this directive would be for developers to explicitly say 'This mutation may cause a conflict on this particular type'. The name here is generic enough that if we need to add additional logic in there to handle conflicts in the future then that would be possible.